### PR TITLE
TATAU-228 Copy data to TaskDeclaration from Model and Dataset

### PR DESCRIPTION
**Pull request by makar21**
TATAU-228

and:

- added timeout to bdb requests
- wrap all requests to parity blockchain if FREE_MODE==True

### CODE REVIEW

File: tatau_core/models/task.py [hunk 1]
```diff
@@ -63,6 +64,16 @@ class Type:
...
 
+    # this field is using as initial weights and train result
+    weights_ipfs = fields.CharField(required=False)
+
+    # in case of horovod will be copied from master worker
+    train_loss = fields.FloatField(required=False)
+    train_accuracy = fields.FloatField(required=False)
+
+    eval_loss = fields.FloatField(required=False)
+    eval_accuracy = fields.FloatField(required=False)
```
Comment (Yulya) [inline tatau_core/models/task.py]: producer_node line 698: change loss, accuracy to eval_loss, eval_accuracy
Look at other usages also, there might be more

Comment (makar21) [reply tatau_core/models/task.py]: You were right. Thanks.
Now it fixed.

File: manage-tasks.py [hunk 2]
```diff
@@ -31,7 +31,7 @@ def create_model(producer, model_name, model_code_path):
     )
 
 
-def create_environment(producer, name, image_name, path_to_requirements):
+def create_environment(producer, name, image_name):
```
Comment (27tech) [inline manage-tasks.py]: Add doc

File: tatau_core/models/dataset.py [hunk 3]
```diff
@@ -27,63 +27,36 @@ class Type:
     train = fields.EncryptedCharField(required=False)
     test = fields.EncryptedCharField(required=False)
 
-    S3_train_ds_uploaded = fields.IntegerField(initial=0)
-    S3_test_ds_uploaded = fields.IntegerField(initial=0)
-
-    def get_chunk_for_estimate(self):
-        ipfs_dir = Directory(self.train)
+    @classmethod
+    def get_chunk_for_estimate(cls, ipfs_dir, dataset_type):
```
Comment (27tech) [inline tatau_core/models/dataset.py]: add doc

File: tatau_core/models/environment.py [hunk 4]
```diff
@@ -1,15 +1,8 @@
 from tatau_core.db import models, fields
-from tatau_core.utils.ipfs import IPFS
 
 
 class Environment(models.Model):
-    name = fields.CharField(immutable=True)
-    image = fields.CharField(immutable=True)
-    requirements = fields.CharField(immutable=True, null=True)
+    name = fields.EncryptedCharField()
```
Comment (27tech) [inline tatau_core/models/environment.py]:
```suggestion
    #: Docker image name
    name = fields.EncryptedCharField()
```

File: tatau_core/models/estimation.py [hunk 5]
```diff
@@ -13,6 +13,7 @@ class EstimationData(models.Model):
 
     chunk_ipfs = fields.EncryptedCharField()
     model_code_ipfs = fields.EncryptedCharField()
+    environment_image_name = fields.EncryptedCharField()
```
Comment (27tech) [inline tatau_core/models/estimation.py]:
```suggestion
    environment_image = fields.EncryptedCharField()
```

File: tatau_core/models/task.py [hunk 6]
```diff
@@ -43,13 +43,14 @@ class Type:
...
-    accuracy = fields.FloatField(required=False)
-
-    train_loss = fields.FloatField(required=False)
-    train_accuracy = fields.FloatField(required=False)
+    # task parameters, copy from dataset and model assets
+    model_code_ipfs = fields.EncryptedCharField(immutable=True)
+    dataset_type = fields.EncryptedCharField(immutable=True)
+    dataset_train_ipfs = fields.EncryptedCharField(immutable=True)
+    dataset_test_ipfs = fields.EncryptedCharField(immutable=True)
+    environment_image_name = fields.EncryptedCharField(immutable=True)
```
Comment (27tech) [inline tatau_core/models/task.py]:
```suggestion
    environment_image = fields.EncryptedCharField(immutable=True)
```

File: tatau_core/models/train.py [hunk 7]
```diff
@@ -17,9 +17,8 @@ class TrainData(models.Model):
 
     dataset_type = fields.EncryptedCharField()
     train_chunks_ipfs = fields.EncryptedJsonField()
-
-    # data for evaluation
     test_chunks_ipfs = fields.EncryptedJsonField()
+    environment_image_name = fields.EncryptedCharField()
```
Comment (27tech) [inline tatau_core/models/train.py]:
```suggestion
    environment_image = fields.EncryptedCharField()
```

File: tatau_core/models/verification.py [hunk 8]
```diff
@@ -13,6 +13,7 @@ class VerificationData(models.Model):
     dataset_type = fields.EncryptedCharField(immutable=True)
     test_dir_ipfs = fields.EncryptedCharField(immutable=True)
     model_code_ipfs = fields.EncryptedCharField(immutable=True)
+    environment_image_name = fields.EncryptedCharField(immutable=True)
```
Comment (27tech) [inline tatau_core/models/verification.py]:
```suggestion
    environment_image = fields.EncryptedCharField(immutable=True)
```

File: tatau_core/node/producer/producer_node.py [hunk 9]
```diff
@@ -63,7 +60,7 @@ def _assign_estimate_data(self, task_declaration: TaskDeclaration):
             assert False and 'Check query!'
 
         if len(timeout_estimation_assignments):
-            # its reassign
+            # this is reassign
```
Comment (27tech) [inline tatau_core/node/producer/producer_node.py]: Need more details

File: tatau_core/node/producer/producer_node.py [hunk 10]
```diff
@@ -93,6 +90,7 @@ def _assign_estimate_data(self, task_declaration: TaskDeclaration):
                 estimation_data = EstimationData.create(
                     db=self.db,
                     encryption=self.encryption,
+                    environment_image_name=task_declaration.environment_image_name,
```
Comment (27tech) [inline tatau_core/node/producer/producer_node.py]:
```suggestion
                    environment_image=task_declaration.environment_image_name,
```

File: tatau_core/node/producer/producer_node.py [hunk 11]
```diff
@@ -623,9 +627,10 @@ def _reassign_verification_data(self, task_declaration: TaskDeclaration):
                 verification_assignment_id=va.asset_id,
                 # share data with verifier
                 public_key=va.verifier.enc_key,
-                dataset_type=task_declaration.dataset.type,
-                test_dir_ipfs=task_declaration.dataset.test,
-                model_code_ipfs=task_declaration.train_model.code_ipfs,
+                dataset_type=task_declaration.dataset_type,
+                test_dir_ipfs=task_declaration.dataset_test_ipfs,
+                model_code_ipfs=task_declaration.model_code_ipfs,
+                environment_image_name=task_declaration.environment_image_name,
```
Comment (27tech) [inline tatau_core/node/producer/producer_node.py]:
```suggestion
                environment_image=task_declaration.environment_image_name,
```

File: tatau_core/sessions/estimation.py [hunk 12]
```diff
@@ -42,10 +42,7 @@ def process_assignment(self, assignment: EstimationAssignment):
 
         assignment.estimation_result.progress = 20.0
         assignment.estimation_result.save()
-
-        task_declaration = assignment.task_declaration
-        environment = task_declaration.environment
-        image = environment.image
+        image = assignment.estimation_data.environment_image_name
```
Comment (27tech) [inline tatau_core/sessions/estimation.py]:
```suggestion
        image = assignment.estimation_data.environment_image
```

File: tatau_core/sessions/eval_train.py [hunk 13]
```diff
@@ -64,7 +64,7 @@ def process_assignment(self, assignment: TaskAssignment, *args, **kwargs):
             weights_ipfs=weights_ipfs,
             test_chunks_ipfs=assignment.train_data.test_chunks_ipfs,
             dataset_type=assignment.train_data.dataset_type,
-            image=assignment.task_declaration.environment.image
+            image=assignment.train_data.environment_image_name
```
Comment (27tech) [inline tatau_core/sessions/eval_train.py]:
```suggestion
            image=assignment.train_data.environment_image
```

File: tatau_core/sessions/eval_verification.py [hunk 14]
```diff
@@ -33,7 +33,7 @@ def process_assignment(self, assignment: VerificationAssignment, *args, **kwargs
             batch_size=assignment.task_declaration.batch_size,
             test_chunks_ipfs=test_chunks_ipfs,
             dataset_type=assignment.verification_data.dataset_type,
-            image=assignment.task_declaration.environment.image
+            image=assignment.verification_data.environment_image_name
```
Comment (27tech) [inline tatau_core/sessions/eval_verification.py]:
```suggestion
            image=assignment.verification_data.environment_image
```

File: tatau_core/sessions/hvd_train.py [hunk 15]
```diff
@@ -172,7 +171,7 @@ def process_assignment(self, assignment: TaskAssignment, *args, **kwargs):
 
         args_list += [str(a) for a in args]
         logger.info('Run master container')
-        logs = self._run(args_list, image=image)
+        logs = self._run(args_list, image=assignment.train_data.environment_image_name)
```
Comment (27tech) [inline tatau_core/sessions/hvd_train.py]:
```suggestion
        logs = self._run(args_list, image=assignment.train_data.environment_image)
```

File: tatau_core/sessions/hvd_train_slave.py [hunk 16]
```diff
@@ -78,7 +73,7 @@ def process_assignment(self, assignment: TaskAssignment, *args, **kwargs):
         ssh.add_private_key(assignment.worker.encryption.private_key.exportKey('PEM'), path_suffix=self.uuid)
 
         logger.info('Slave container run')
-        self._run(args_list, image=image, detach=True)
+        self._run(args_list, image=assignment.train_data.environment_image_name, detach=True)
```
Comment (27tech) [inline tatau_core/sessions/hvd_train_slave.py]:
```suggestion
        self._run(args_list, image=assignment.train_data.environment_image, detach=True)
```

File: tatau_core/sessions/summarize.py [hunk 17]
```diff
@@ -50,7 +50,7 @@ def process_assignment(self, assignment: VerificationAssignment, *args, **kwargs
         self.model_path = downloader.resolve_path('model.py')
         self.results_list = downloaded_results
 
-        self._run(self.start_session_args, image=assignment.task_declaration.environment.image)
+        self._run(self.start_session_args, image=assignment.verification_data.environment_image_name)
```
Comment (27tech) [inline tatau_core/sessions/summarize.py]:
```suggestion
        self._run(self.start_session_args, image=assignment.verification_data.environment_image)
```

File: tatau_core/sessions/train.py [hunk 18]
```diff
@@ -56,7 +56,7 @@ def process_assignment(self, assignment: TaskAssignment, *args, **kwargs):
         args_list += [str(batch_size), str(epochs), str(assignment.train_data.current_iteration)]
         args_list += [str(a) for a in args]
 
-        self._run(args_list, image=assignment.task_declaration.environment.image)
+        self._run(args_list, image=assignment.train_data.environment_image_name)
```
Comment (27tech) [inline tatau_core/sessions/train.py]:
```suggestion
        self._run(args_list, image=assignment.train_data.environment_image)
```

File: tatau_core/settings/__init__.py [hunk 19]
```diff
@@ -42,6 +42,7 @@
 VALID_TRANSACTIONS_STREAM_URL = 'ws://{}:9985/api/v1/streams/valid_transactions'.format(BDB_HOST)
 
 BDB_ROOT_URL = 'http://{}:9984'.format(BDB_HOST)
+BDB_COMMUNICATION_TIMEOUT = int(os.getenv('BDB_COMMUNICATION_TIMEOUT', 300))
```
Comment (27tech) [inline tatau_core/settings/__init__.py]:
```suggestion
BDB_TIMEOUT = int(os.getenv('BDB_COMMUNICATION_TIMEOUT', 300))
```
